mage_ai/orchestration/db/models/oauth.py

```diff
@@ -89,6 +123,140 @@ def git_settings(self) -> Union[Dict, None]:
         preferences = self.preferences or dict()
         return preferences.get(get_repo_path(), {}).get('git_settings')
 
+    @classmethod
+    @safe_db_query
+    def batch_update_user_roles(self):
+        for user in User.query.all():
+            roles_new = []
+            if user._owner:
+                roles_new = [Role.get_role('Owner')]
+            elif user.roles and user.roles & 1 != 0:
+                roles_new = [Role.get_role('Admin')]
+            elif user.roles and user.roles & 2 != 0:
+                roles_new = [Role.get_role('Editor')]
+            elif user.roles and user.roles & 4 != 0:
+                roles_new = [Role.get_role('Viewer')]
+            user.roles_new = roles_new
+        db_connection.session.commit()
+
+
+class Role(BaseModel):
+    name = Column(String(255))
+    permissions = relationship('Permission', back_populates='role')
+    users = relationship('User', secondary='user_role', back_populates='roles_new')
+
+    @classmethod
+    @safe_db_query
+    def create_default_roles(self):
+        Permission.create_default_global_permissions()
+        mapping = {
+            'Owner': Permission.Access.OWNER,
+            'Admin': Permission.Access.ADMIN,
+            'Editor': Permission.Access.EDITOR,
+            'Viewer': Permission.Access.VIEWER,
+        }
+        for name, access in mapping.items():
+            role = self.query.filter(self.name == name).first()
+            if not role:
+                self.create(
+                    name=name,
+                    permissions=[
+                        Permission.query.filter(
+                            Permission.entity == Permission.Entity.GLOBAL,
+                            Permission.access == access,
+                        ).first()
+                    ],
+                    commit=False,
+                )
+        db_connection.session.commit()
+
+    @classmethod
+    @safe_db_query
+    def get_role(self, name) -> 'Role':
+        return Role.query.filter(Role.name == name).first()
+
+    def get_access(
+        self,
+        entity: Union['Permission.Entity', None] = None,
+        entity_id: Union[str, None] = None,
+    ) -> int:
+        permissions = []
+        if entity is None:
+            permissions.extend(self.permissions)
+        else:
+            entity_permissions = list(filter(
+                lambda perm: perm.entity == entity and
+                (entity_id is None or perm.entity_id == entity_id),
+                self.permissions,
+            ))
+            if entity_permissions:
+                permissions.extend(entity_permissions)
+
+        access = 0
+        if permissions:
+            for permission in permissions:
+                access = access | permission.access
+            return access
+        else:
+            return self.get_parent_access(entity)
```

Comment:
Actually, i feel there're some corner cases here, which is not an issue for now but might be an issue for later (custom role).
If a role has 2 permissions (view pipeline 1, edit project), then when we get_access for pipeline 1, it'll only return view permission. But it actually also has edit access for the pipeline.

You can add a TODO here to handle role of permissions with different entity types.

Reply:
Hmm yeah I guess I was thinking if they have a permission on a more specific entity, that would override the "parent" entity permission. I could definitely see other corner cases though, I'll add a TODO